proposed/2021/ProjectUpdateEventsInVisualStudio.md

```diff
@@ -0,0 +1,152 @@
+In addition, NuGet will provide a solution restore start/end event which will include a list of PackageReference projects only.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+With SDK based projects, NuGet is tightly integrated with solution load & design time builds.
+3 files are of interest for design time builds, `project.assets.json`, `projectName.projectExt.nuget.g.[props/targets]`.
+When a restore happens, one or all of these files may be touched.
+Given that these files are edited in quick succession, normally you'd end up getting only 1 design time build, but in some scenarios, the writes may take longer due to ambient reasons and cause extra design time builds.
+
+`packages.config` and `PackageReference` projects differ fundamentally in the way they deal with package changes.
+In `packages.config` projects, NuGet writes to the *project* files, and NuGet references are equivalent to any other new references.
+In `PackageReference` projects, the contract is expressed through the assets file. An assets file change may mean that 10 packages have changes.
+Components such as test explorer, need to know when packages for a project have changed to scan for adapters.
+For `packages.config` projects, one can use [`IVsPackageInstallerEvents`](https://docs.microsoft.com/en-us/nuget/visual-studio-extensibility/nuget-api-in-visual-studio#ivspackageinstallerevents-interface). There's no PackageReference equivalent.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+Given that this is an extensibility feature, there's no NuGet experiences impact.
+The functional impact matches what the consumers of this API use it for.
+
+- For SDK based projects, the project-system will pause eager reevaluation while the writes of the `project.assets.json` and `projectName.projectExt.nuget.g.[props/targets]` are ongoing.
+- The test explorer will be able to refer to listen to a single API for project updates.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+The API and the shape is fairly self explanatory.
+
+```cs
+namespace NuGet.SolutionRestoreManager
+{
+    /// <summary>
+    /// NuGet project update events.
+    /// Architecturally, packages.config and PackageReference projects differ in the way package updates are processed.
+    /// This interface is meant to provide a single API of interest for components wanting to listen to *all* project updates by NuGet.
+    /// </summary>
+    [ComImport]
+    [Guid("30CDDD0A-6901-482D-8CEF-6D798F1A99FC")]
+    public interface IVsNuGetProjectUpdateEvents
```

Comment:
how do we access this component? is the implementation free threaded & do we expect events to be fired in a thread pool thread?

Reply:
NuGet exposes services through the NuGet.VisualStudio assembly. You should be able to consume IVsNuGetProjectUpdateEvents via MEF. 

The PackageReference events (only ones you care about), are going to be fired from a threadpool thread. 

I'll need to verify how the packages.config ones are fired today.